moshi/src/main/java/com/squareup/moshi/JsonValueReader.java

```diff
@@ -374,4 +393,32 @@ private void remove() {
       }
     }
   }
+
+  static final class JsonIterator implements Iterator<Object>, Cloneable {
+    final Token endToken;
+    final Object[] list;
+    int next;
+
+    JsonIterator(Token endToken, Object[] list, int next) {
+      this.endToken = endToken;
+      this.list = list;
+      this.next = next;
+    }
+
+    @Override public boolean hasNext() {
+      return next < list.length;
+    }
+
+    @Override public Object next() {
+      return list[next++];
+    }
+
+    @Override public void remove() {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override protected JsonIterator clone() {
+      return new JsonIterator(endToken, list, next);
```

Comment:
worth a comment to state that the `list` doesn't need a clone?

Reply:
Renamed to `array`!

Reply:
Fixed.